File: day23.py
#!/opt/local/bin/python

class Game:
    """Implements the crab cup game."""
    def __init__(self, cups):
        self.next = [0]*(1+len(cups))
        first = cups[0]
        n = self.next
        for i in range(len(cups)-1):
            n[cups[i]] = cups[i+1]
        n[cups[-1]] = first
        self.curr = first
        self.max = len(cups)
    def play(self, rounds):
        """Plays the given number of rounds of the cup game."""
        n, curr = self.next, self.curr # object dereferences are expensive
        for _ in range(rounds):
            a = n[curr]
            b = n[a]
            c = n[b]
            d = curr-1
            if d < 1:
                d = self.max
            while (a == d or b == d or c == d):
                d -= 1
                if d < 1:
                    d = self.max
            e = n[c]
            n[curr] = e
            curr = e
            n[c] = n[d]
            n[d] = a
        self.curr = curr
    def after(self, l, n = -1):
        """Prints the n cups clockwise from the cup labeled l
        (or all cups (except l) if n < 0).
        """
        p = l
        res = []
        if n < 0:
            n = self.max-1
        for _ in range(n):
            p = self.next[p]
            res.append(p)
        return res

def part1():
    cups = "653427918"
    game = Game([int(x) for x in cups])
    game.play(100)
    return "".join([str(x) for x in game.after(1)])

def part2():
    cups = "653427918"
    n, rounds = 1000*1000, 10*1000*1000
    c = [int(x) for x in cups] + [x for x in range(10,n+1)]
    game = Game(c)
    game.play(rounds)
    res = game.after(1,2)
    return res[0] * res[1]

print(part1())
print(part2())

File: day23.cc
#include <iostream>
#include <string>
#include <vector>

class Game
{
public:
    Game(std::vector<int> cups, int maxval = -1)
    {
        if (maxval < static_cast<int>(cups.size())) {
            maxval = cups.size();
        }
        max = maxval;
        curr = cups[0];
        next = new int[max+1];
        int i = 0;
        for (; i < cups.size()-1; i++) {
            next[cups[i]] = cups[i+1];
        }
        if (max == cups.size()) {
            next[cups[i]] = curr;
            return;
        }
        next[cups[i]] = cups.size()+1;
        for (i = cups.size()+1; i < max; i++) {
            next[i] = i+1;
        }
        next[max] = curr;
    }
    ~Game() {
        delete[] next;
    }
    void play(int n) {
        for (int i = 0; i < n; i++) {
            int a = next[curr], b = next[a], c = next[b];
            int d = curr-1;
            if (d < 1) {
                d = max;
            }
            while (a == d || b == d || c == d) {
                d--;
                if (d < 1) {
                    d = max;
                }
            }
            next[curr] = next[c];
            curr = next[c];
            next[c] = next[d];
            next[d] = a;
        }
    }
    std::vector<int> after(int l, int n = -1) {
        if (n < 0)
            n = max-1;
        std::vector<int> res(n);
        for (int i = 0; i < n; i++) {
            l = next[l];
            res[i] = l;
        }
        return res;
    }

private:
    int max;
    int curr;
    int* next;
};

std::string part1()
{
    std::vector<int> v = { 6,5,3,4,2,7,9,1,8 };
    Game game(v);
    game.play(100);
    std::vector<int> res = game.after(1);
    std::string str;
    for (int i = 0; i < res.size(); i++) {
        str += std::to_string(res[i]);
    }
    return str;
}

int64_t part2()
{
    std::vector<int> v = { 6,5,3,4,2,7,9,1,8 };
    Game game(v, 1000*1000);
    game.play(10*1000*1000);
    std::vector<int> res = game.after(1,2);
    int64_t p = (int64_t)res[0] * res[1];
    return p;
}

int main()
{
    std::cout << part1() << std::endl;
    std::cout << part2() << std::endl;
    return 0;
}
